cmd/loadgenerator/main.go

```diff
@@ -0,0 +1,222 @@
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Cannot recover key seed from mnemonic: %v\n", err)
+		os.Exit(1)
+	}
+
+	var seed crypto.Seed
+	copy(seed[:], seedbytes)
+	return seed
+}
+
+func main() {
+	var cfg config
+	var err error
+	if cfg, err = loadConfig(); err != nil {
+		fmt.Fprintf(os.Stderr, "unable to load config : %v\n", err)
+		os.Exit(1)
+	}
+	fmt.Printf("Configuration file loaded successfully.\n")
+
+	seed := loadMnemonic(cfg.AccountMnemonic)
+	privateKey := crypto.GenerateSignatureSecrets(seed)
+	publicKey := basics.Address(privateKey.SignatureVerifier)
+
+	fmt.Printf("Spending account public key : %v\n", publicKey.String())
+
+	err = spendLoop(cfg, privateKey, publicKey)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "spend loop error : %v\n", err)
+		os.Exit(1)
+	}
+
+	os.Exit(0)
+}
+
+func isSpendRound(cfg config, round uint64) bool {
+	return cfg.RoundModulator == 0 || ((round+cfg.RoundOffset)%cfg.RoundModulator == 0)
+}
+
+func nextSpendRound(cfg config, round uint64) uint64 {
+	if cfg.RoundModulator == 0 {
+		return round
+	}
+	return ((round+cfg.RoundOffset)/cfg.RoundModulator)*cfg.RoundModulator + cfg.RoundModulator
+}
+
+func spendLoop(cfg config, privateKey *crypto.SignatureSecrets, publicKey basics.Address) (err error) {
+	restClient := client.MakeRestClient(*cfg.ClientURL, cfg.APIToken)
+
+	for {
+		waitForSpendingRound(restClient, cfg)
+		queueFull := generateTransactions(restClient, cfg, privateKey, publicKey)
+		if queueFull {
+			waitForNonSpendingRound(restClient, cfg)
+		}
+	}
+}
+
+func waitForNonSpendingRound(restClient client.RestClient, cfg config) {
+	var nodeStatus generatedV2.NodeStatusResponse
+	var err error
+	for {
+		nodeStatus, err = restClient.Status()
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "unable to check status : %v", err)
+			time.Sleep(1 * time.Second)
+			continue
+		}
+		if !isSpendRound(cfg, nodeStatus.LastRound) {
+			return
+		}
+	waitForNextBlock:
+		// wait for the next round.
+		nodeStatus, err = restClient.WaitForBlock(basics.Round(nodeStatus.LastRound))
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "unable to wait for next round node status : %v", err)
+			time.Sleep(1 * time.Second)
+			continue
+		}
+		if !isSpendRound(cfg, nodeStatus.LastRound) {
+			return
+		}
+		goto waitForNextBlock
+	}
+}
+
+func waitForSpendingRound(restClient client.RestClient, cfg config) {
+	var nodeStatus generatedV2.NodeStatusResponse
+	var err error
+	for {
+		nodeStatus, err = restClient.Status()
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "unable to check status : %v", err)
+			time.Sleep(1 * time.Second)
+			continue
+		}
+		if isSpendRound(cfg, nodeStatus.LastRound) {
+			// time to send transactions.
+			return
+		}
+		fmt.Printf("Current round %d, waiting for spending round %d\n", nodeStatus.LastRound, nextSpendRound(cfg, nodeStatus.LastRound))
+	waitForNextBlock:
+		// wait for the next round.
+		nodeStatus, err = restClient.WaitForBlock(basics.Round(nodeStatus.LastRound))
```

Comment:
What am I missing here? It this waiting for the current round? Isn't this trivial?
Did you mean waiting for nextRpendRound?

Reply:
nope; this one was abit tricky for me too. If you'll look on the server side implementation of WaitForBlock you'll see that it's waiting for the round+1 to get committed. 
while this isn't trivial, changing this behavior at the source would cause more failures than we want to have; for sure not something to be done in the context of this solution.

Reply:
Then why not wait for next spending round - 1?

Reply:
Waiting for "next spending round - 1" can take a long time to reach. The server is limiting the request to 1 minute anyway. In general, I don't like to leave the server holding a connection for a long time, so keeping it for a single round seems fair.